fuel/datasets/hdf5.py

```diff
@@ -64,3 +67,124 @@ def get_data(self, state=None, request=None):
+    subset : slice, optional
+        A slice of data *within the context of the split* to use. Defaults
+        to `None`, in which case the whole split is used. **Note:
+        at the moment, `slice.step` must be either 1 or `None`.**
+    load_in_memory : bool, optional
+        Whether to load the data in main memory. Defaults to `False`.
+    driver : str, optional
+        Low-level driver to use. Defaults to `None`. See h5py
+        documentation for a complete list of available options.
+
+    """
+    ref_counts = dict()
+
+    def __init__(self, path, which_set=None, subset=None, load_in_memory=False,
+                 driver=None, **kwargs):
+        self.path = path
+        self.which_set = which_set
+        self.subset = subset if subset else slice(None, None, None)
+        self.load_in_memory = load_in_memory
+        self.driver = driver
+        self.load()
+        super(H5PYDataset, self).__init__(**kwargs)
+
+    def _get_file_id(self):
+        try:
+            return next(
+                filter(lambda x: x.name == self.path, self.ref_counts.keys()))
+        except StopIteration:
+            return self.path
+
+    def load(self):
+        handle = self._out_of_memory_open()
+        self.provides_sources = list(handle.keys())
+        shapes = [data_source.shape for data_source in handle.values()]
+        if any(s[0] != shapes[0][0] for s in shapes):
+            raise ValueError("sources have different lengths")
+        if self.subset.step not in (1, None):
+            raise ValueError("subset.step must be either 1 or None")
+        start, stop = (handle.attrs[self.which_set] if self.which_set
+                       else (0, shapes[0][0]))
+        self.subset = slice(
+            start if self.subset.start is None else self.subset.start,
+            stop if self.subset.stop is None else self.subset.stop,
+            self.subset.step)
+        self.num_examples = self.subset.stop - self.subset.start
+        self.data_sources = ([data_source[self.subset] for
+                              data_source in handle.values()]
+                             if self.load_in_memory else None)
+        self._out_of_memory_close(handle)
+
+    def open(self):
+        return None if self.load_in_memory else self._out_of_memory_open()
+
+    def _out_of_memory_open(self):
+        file_id = self._get_file_id()
+        state = h5py.File(name=file_id, mode="r", driver=self.driver)
+        self.ref_counts[state.id] = self.ref_counts.get(state.id, 0) + 1
+        return state
+
+    def close(self, state):
+        if not self.load_in_memory:
+            self._out_of_memory_close(state)
+
+    def _out_of_memory_close(self, state):
+        self.ref_counts[state.id] -= 1
+        if not self.ref_counts[state.id]:
+            del self.ref_counts[state.id]
+            state.close()
+
+    def get_data(self, state=None, request=None):
+        get_data_method = (self._in_memory_get_data if self.load_in_memory
```

Comment:
This is just a matter of preference I guess, but some people would argue the ternary operator isn't very pythonic. Sometimes I think it's useful, but here I would go for the less dense
```python
if self.load_in_memory:
    return self._in_memory_get_data(*args, **kwargs)
else:
    return self._out_of_memory_get_data(*args, **kwargs)
```